pkg/apis/traceflow/v1/types.go

```diff
@@ -0,0 +1,155 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type DropReason string
+
+type Phase string
+
+const (
+	INITIAL Phase = "INITIAL"
+	RUNNING Phase = "RUNNING"
+	SUCCESS Phase = "SUCCESS"
+	TIMEOUT Phase = "TIMEOUT"
+	ERROR   Phase = "ERROR"
+)
+
+type Component string
+
+const (
+	SPOOFGUARD    Component = "SPOOFGUARD"
+	LB            Component = "LB"
+	ROUTING       Component = "ROUTING"
+	NETWORKPOLICY Component = "NETWORKPOLICY"
+	FORWARDING    Component = "FORWARDING"
+)
+
+type Action string
+
+const (
+	DELIVERED Action = "DELIVERED"
+	RECEIVED  Action = "RECEIVED"
+	FORWARDED Action = "FORWARDED"
+	DROPPED   Action = "DROPPED"
+)
+
+type PacketTransportType string
+
+const (
+	UNICAST PacketTransportType = "UNICAST"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Source      `json:"source"`
+	Destination `json:"destination"`
+	RoundID     string `json:"roundID,omitempty"`
+
+	Packet `json:"packet"`
+	Status `json:"status"`
+}
+
+type Source struct {
+	Namespace string `json:"namespace,omitempty"`
+	Pod       string `json:"pod,omitempty"`
+}
+
+type Destination struct {
+	Namespace string `json:"namespace,omitempty"`
+	Pod       string `json:"pod,omitempty"`
+	Service   string `json:"service,omitempty"`
+	IP        string `json:"IP,omitempty"`
+}
+
+type IPHeader struct {
+	SrcIP    string `json:"srcIP,omitempty"`
+	Protocol int32  `json:"protocol,omitempty"`
+	TTL      int32  `json:"ttl,omitempty"`
+	Flags    int32  `json:"flags,omitempty"`
+}
+
+type TransportHeader struct {
+	ICMPEchoRequestHeader `json:"icmpEchoRequestHeader"`
+	UDPHeader             `json:"udpHeader"`
+	TCPHeader             `json:"tcpHeader"`
+}
+
+type ICMPEchoRequestHeader struct {
+	ID       int32 `json:"id,omitempty"`
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+type UDPHeader struct {
+	SrcPort int32 `json:"srcPort,omitempty"`
+	DstPort int32 `json:"dstPort,omitempty"`
+}
+
+type TCPHeader struct {
+	SrcPort int32 `json:"srcPort,omitempty"`
+	DstPort int32 `json:"dstPort,omitempty"`
+	Flags   int32 `json:"flags,omitempty"`
+}
+
+type Packet struct {
+	TransportType PacketTransportType `json:"transportType,omitempty"`
+	PayloadString string              `json:"payloadString,omitempty"`
+	PayloadSize   int32               `json:"payloadSize,omitempty"`
+
+	IPHeader        `json:"ipHeader"`
+	TransportHeader `json:"transportHeader"`
+}
+
+type Status struct {
+	Phase        Phase         `json:"phase,omitempty"`
+	CrossNodeTag uint8         `json:"crossNodeTag,omitempty"` // unique ID for traceflow
+	NodeSender   []Observation `json:"nodeSender,omitempty"`
+	NodeReceiver []Observation `json:"nodeReceiver,omitempty"`
+}
+
+type Observation struct {
+	Component       Component  `json:"component,omitempty"`
+	ComponentName   string     `json:"componentName,omitempty"`
```

Comment:
Why we need this field, besides "Component"?

Reply:
Thought over. I guess let us remove it for now. Feel "ingress"/"egress" for NetworkPolicy is not very useful.

Reply:
Removed for now. Will add it back if needed.

Reply:
@jianjuns I wonder if you agree to keep it by changing its name. For example, "SubComponent" or "ComponentInfo". This "ComponentName" functions as an extension of "Component".

Reply:
I feel Ingress/Egress is not very useful, as it can be figured out from the observation already? Do you have other use cases?
ComponentInfo sounds better to me though.

Reply:
Currently we also fill RuleTable in FORWARDING component.

Reply:
Sorry, what we fill for FORWARDING? OVS table name?

Reply:
Yes. Such as "L2 Forwarding Output" or "SpoofGuard"

Reply:
I thought SpoofGuard is a separate component? Do you have examples for one component, you want to add more information?

Reply:
"Classifier" for component:"FORWARDING", action:"RECEIVED", this case is used when packet coming from tunnel on ReceiverNode.

Reply:
Ok. Then maybe let us rename it ComponentInformation, but let us be careful not to add useless extra information.

Reply:
Got it. Thanks.

Reply:
Use ComponentInfo now.